project1/thread_pool/task.h

```diff
@@ -0,0 +1,60 @@
+#ifndef PROJECT1_TASK_H
+#define PROJECT1_TASK_H
+
+#include <cstdint>
+
+namespace thread_pool {
+
+/**
+ * @brief Represents a task that can be run in a tread pool.
+ */
+class Task {
+ public:
+  /// Represents an opaque identifier for a task.
+  using Handle = uint32_t;
+
+  /**
+   * @brief Represents the status of a task.
+   */
+  enum class Status {
+    /// Task is running normally.
+    RUNNING,
+    /// Task has completed normally.
+    DONE,
+    /// Task has failed.
+    FAILED,
+  };
+
+  /**
+   * @brief Performs any necessary one-time setup when the task starts.
+   * @return The updated status for this task.
+   */
+  virtual Status SetUp() = 0;
+
+  /**
+   * @brief Runs one iteration of the task main loop.
+   * @note The task cannot be cancelled in the middle of this function. However,
+   *    you should assume that cancellation is possible between calls.
+   * @return The updated status for this task.
+   */
+  virtual Status RunAtomic() = 0;
```

Comment:
I'm not sure what you mean by `task main loop`?

Reply:
After reading over this and some of your comments below, I think there might be some confusion on how this is intended to be used. I think this will become much more clear once I submit the actual implementation and unit tests.

The `Task` class is not meant to be used directly. New tasks are meant to be created by subclassing `Task` and implementing `RunAtomic` so that it does whatever you want your task to do. The terminology stems from the notion that most tasks are expected to consist of some loop (i.e. a "main loop") that gets run repeatedly. For instance, you might have a task that listens on a socket, handles a command that it received, and then goes back to listening on the socket for the next command.

I recognize that this implementation strategy is a little confusing compared to just having a `Run` method that runs the entire task in a monolithic block. It is done this way in order to implement task cancellation without having to have any (explicit) buy-in from specific task implementations.